app/scripts/controllers/network/provider-api-tests/shared-tests.js

```diff
@@ -129,6 +138,377 @@ export function testsForRpcMethodAssumingNoBlockParam(method) {
+      comms.mockNextBlockTrackerRequest();
+      comms.mockInfuraRpcCall({
+        request,
+        response: {
+          id: 12345,
+          jsonrpc: '2.0',
+          error: 'some error',
+          httpStatus: 420,
+        },
+      });
+      const promiseForResult = withInfuraClient(async ({ makeRpcCall }) =>
+        makeRpcCall(request),
+      );
+
+      await expect(promiseForResult).rejects.toThrow(
+        '{"id":12345,"jsonrpc":"2.0","error":"some error"}',
+      );
+    });
+  });
+
+  [503, 504].forEach((httpStatus) => {
+    it(`retries the request to Infura up to 5 times if it returns a ${httpStatus} response, returning the successful result if there is one on the 5th try`, async () => {
+      await withMockedInfuraCommunications(async (comms) => {
+        const request = { method };
+
+        // The first time a block-cacheable request is made, the latest block
+        // number is retrieved through the block tracker first. It doesn't
+        // matter what this is — it's just used as a cache key.
+        comms.mockNextBlockTrackerRequest();
+        // Here we have the request fail for the first 4 tries, then succeed
+        // on the 5th try.
+        comms.mockInfuraRpcCall({
+          request,
+          response: {
+            error: 'Some error',
+            httpStatus,
+          },
+          times: 4,
+        });
+        comms.mockInfuraRpcCall({
+          request,
+          response: {
+            result: 'the result',
+            httpStatus: 200,
+          },
+        });
+        const result = await withInfuraClient(
+          async ({ makeRpcCall, clock }) => {
+            // Note that we have to manually capture the value that the
+            // promise that `makeRpcCall` returns when it resolves or is
+            // rejected using `then`/`catch` instead of `try`/`catch` because
+            // the resolution/rejection occurs out of band. To be more
+            // specific, the promise will be fulfilled while the `while` loop
+            // below is running. This means that if we did not attach `catch`
+            // to the promise, Node would complain that the promise had a
+            // unhandled rejection. So we attach `catch` to capture the
+            // rejection and we also attach `then` to capture the resolution
+            // and check them after the loop ends.
+
+            let capturedResolutionValue;
+            let capturedRejectionValue;
+            makeRpcCall(request)
+              .then((resolutionValue) => {
+                capturedResolutionValue = resolutionValue;
+              })
+              .catch((rejectionValue) => {
+                capturedRejectionValue = rejectionValue;
+              });
+
+            while (nock.pendingMocks().length > 0) {
+              clock.runAll();
+              await new Promise((resolve) =>
+                originalSetTimeout(resolve, TIME_TO_WAIT_FOR_NEXT_INFURA_RETRY),
```

Comment:
If we're using fake timers here, why would we need to use the real `setTimeout`? I thought the fake timers API was supposed to flush everything upon calling `runAll`.

Reply:
The retry logic in the Infura middleware within `createInfuraMiddleware` will attempt to hit Infura. If the request has failed, then it will start a 1 second timer, then retry the request after the timer has completed. This happens up to 5 times.

We are stubbing timers via Jest, so when a timer gets scheduled, we have to manually run that timer. The problem is, we don't know when our timer will get scheduled. We can call `clock.runAll()`, but our timer may have not gotten scheduled yet. We could keep running `clock.runAll()` over and over in a loop, but I found that if we wait a bit before re-running `clock.runAll()` then the tests reliably succeed. Without this bit, they will fail intermittently.

Reply:
Ah I see, makes sense.

Could we use `process.nextTick` here instead? It seems that sinon intentionally omits that from the set of default timers that get stubbed out so that there is a way to handle this situation.

Reply:
Actually scratch that, this doesn't seem well suited to what `process.nextTick` is designed for. I think the solution you have now of capturing the unstubbed `setTimeout` is probably better and less confusing.